core/src/main/java/org/apache/iceberg/ManifestReader.java

```diff
@@ -214,16 +216,31 @@ private void cacheChanges() {
 
     switch (format) {
       case AVRO:
-        AvroIterable<ManifestEntry> reader = Avro.read(file)
-            .project(ManifestEntry.wrapFileSchema(fileProjection.asStruct()))
-            .rename("manifest_entry", GenericManifestEntry.class.getName())
-            .rename("partition", PartitionData.class.getName())
-            .rename("r102", PartitionData.class.getName())
-            .rename("data_file", GenericDataFile.class.getName())
-            .rename("r2", GenericDataFile.class.getName())
-            .classLoader(GenericManifestFile.class.getClassLoader())
-            .reuseContainers()
-            .build();
+        CloseableIterable<ManifestEntry> reader;
+        if (formatVersion < 2) {
+          reader = Avro.read(file)
+              .project(ManifestEntry.wrapFileSchema(fileProjection.asStruct()))
+              .rename("manifest_entry", GenericManifestEntry.class.getName())
+              .rename("partition", PartitionData.class.getName())
+              .rename("r102", PartitionData.class.getName())
+              .rename("data_file", GenericDataFile.class.getName())
+              .rename("r2", GenericDataFile.class.getName())
+              .classLoader(GenericManifestFile.class.getClassLoader())
+              .reuseContainers()
+              .build();
+        } else {
+          AvroIterable<GenericDataFile> files = Avro.read(file)
```

Comment:
Yes. This ensures that the correct schema is used to read a v2 manifest.

Reply:
It is very likely that we will also track this in the manifest list as well, but this is a good work-around to keep the changes separate. And it will help us detect if other metadata is ever wrong.